experimental/transliteration/src/transliterator/replaceable.rs

```diff
@@ -0,0 +1,1040 @@
+            // SAFETY: `len` is guaranteed to be a valid UTF-8 length starting at `forward_cursor`.
+            if KEY_FINISHED {
+                self.post_match_len += len;
+            } else {
+                self.key_match_len += len;
+            }
+            self.forward_cursor += len;
+            true
+        } else {
+            false
+        }
+    }
+
+    fn next_char(&self) -> Option<char> {
+        self.remaining_forward_slice().chars().next()
+    }
+}
+
+// we can always reverse match, no matter if the key has finished matching or not
+impl<'a, 'b, const KEY_FINISHED: bool> Utf8Matcher<Reverse> for RepMatcher<'a, 'b, KEY_FINISHED> {
+    fn cursor(&self) -> usize {
+        self.ante_cursor()
+    }
+
+    fn str_range(&self, range: Range<usize>) -> Option<&str> {
+        self.rep.as_str().get(range)
+    }
+
+    fn is_empty(&self) -> bool {
+        self.ante_cursor() == 0
+    }
+
+    fn match_str(&self, s: &str) -> bool {
+        self.remaining_ante_slice().ends_with(s)
+    }
+
+    fn match_start_anchor(&self) -> bool {
+        self.ante_cursor() == 0
+    }
+
+    fn match_end_anchor(&self) -> bool {
+        self.ante_cursor() == self.rep.content.len()
+    }
+
+    fn consume(&mut self, len: usize) -> bool {
+        if len <= self.ante_cursor() {
+            assert!(self
+                .remaining_ante_slice()
+                .is_char_boundary(self.ante_cursor() - len));
+            // SAFETY: `len` is guaranteed to be a valid UTF-8 length reverse-starting at `ante_cursor()`.
+            self.ante_match_len += len;
+            true
+        } else {
+            false
+        }
+    }
+
+    fn next_char(&self) -> Option<char> {
+        self.remaining_ante_slice().chars().next_back()
+    }
+}
+
+mod sealed {
+    pub(crate) trait Sealed {}
+    impl Sealed for super::Forward {}
+    impl Sealed for super::Reverse {}
+}
+
+/// A forward match.
+///
+/// For example, forward-matching `"aft"` on the input `"previous|after"`, with the cursor at the `'|'`,
+/// would return a match, forward-matching `"fter"` would not.
+pub(super) struct Forward;
+/// A reverse match.
+///
+/// For example, reverse-matching `"ious"` on the input `"previous|after"`, with the cursor at the `'|'`,
+/// would return a match, forward-matching `"prev"` would not.
+pub(super) struct Reverse;
+/// The direction a match can be applied. Used in [`Utf8Matcher`].
+///
+/// See [`Forward`] and [`Reverse`] for implementors.
+pub(super) trait MatchDirection: sealed::Sealed {}
+impl MatchDirection for Forward {}
+impl MatchDirection for Reverse {}
+
+/// Matching functionality on strings. Matching can be done in forward or reverse directions, see [`MatchDirection`].
+///
+/// The used indices in method parameters are all compatible with each other.
+// Thought: I don't think this needs to be called *Utf8* matcher, maybe just Matcheable
+
+pub(super) trait Utf8Matcher<D: MatchDirection>: Debug {
```

Comment:
Question:
Looking for advice for a different API than implementing two traits (`Utf8M<Forward>` and `<Reverse>`) with the same method API surface for the same type. Leads to conflicts that need to be fully qualified such as here (in a follow-up PR): https://github.com/unicode-org/icu4x/pull/3965/files#diff-6dd89dfb5721915aafe24a449bc4d2d499df79d63e2f7687e2eccd9384307cddR44

Maybe just explicit method names? `match_str_forward`, `match_str_reverse`, etc?